README

=======================
 makedumpfile's README
=======================

* REQUIREMENTS
  Please download the following library file and install it exactly as below;
  do NOT use "make install".
  - elfutils-0.144.tar.gz
    The "make install" of elfutils installs some commands (ld, readelf, etc.),
    and compiling problems sometimes happen due to using the installed
    commands. To install only the library & header files, use the following
    method:
     # tar -zxvf elfutils-0.144.tar.gz
     # cd elfutils-0.144
     # ./configure
     # make
     #
     # mkdir /usr/local/include/elfutils/
     # cp ./libdw/libdw.h   /usr/local/include/elfutils/libdw.h
     # cp ./libdw/dwarf.h   /usr/local/include/dwarf.h
     # cp ./libelf/libelf.h /usr/local/include/libelf.h
     # cp ./libelf/gelf.h   /usr/local/include/gelf.h
     #
     # cp ./libelf/libelf.a /usr/local/lib/libelf.a
     # cp ./libdw/libdw.a   /usr/local/lib/libdw.a
     # cp ./libasm/libasm.a /usr/local/lib/libasm.a
     # cp ./libebl/libebl.a /usr/local/lib/libebl.a
     #

* BUILD & INSTALL
  1.Get the latest makedumpfile from the following site:
    https://sourceforge.net/projects/makedumpfile/
  2.Uncompress the tar file:
    # tar -zxvf makedumpfile-x.y.z.tar.gz
  3.Enter the makedumpfile subdirectory:
    # cd makedumpfile-x.y.z
  4.Build, and install:
    # make; make install
  5.Build for a different architecture than the host :
    # make TARGET=<arch> ; make install
    where <arch> is the 'uname -m' of the target architecture. 
    The user has to set the environment variable CC to appropriate
    compiler for the target architecture.
  6.Build with lzo support:
    # make USELZO=on ; make install
    The user has to prepare lzo library.
  7.Build with snappy support:
    # make USESNAPPY=on ; make install
    The user has to prepare snappy library.
  8.Build the extension module for --eppic option.
    # make eppic_makedumpfile.so
    The user has to prepare eppic library from the following site:
    http://code.google.com/p/eppic/

* SUPPORTED KERNELS
  This makedumpfile supports the following kernels.

         |       FLATMEM     |   DISCONTIGMEM    |       SPARSEMEM       
         |-------------------+-------------------+------------------------
   Kernel|    | x86| PPC| PPC|    | x86|    | PPC|    | x86|    | PPC|
  Version| x86| _64|  32|  64| x86| _64|ia64|  64| x86| _64|ia64|  64|s390
  -------+----+----+----+----+----+----+----+----+----+----+----+----+----
  2.6.15 | OK | -- |    | -- | -- | -- | -- | -- | -- | -- | -- | -- |
  2.6.16 | OK | OK |    |    | -- | OK | OK | -- | -- |    | -- |    |
  2.6.17 | OK | OK |    |    | -- | OK | -- | -- | -- | OK | -- |    |
  2.6.18 | OK | OK |    | OK | -- | OK | OK | -- | -- | OK | OK | OK |
  2.6.19 | OK | OK |    | OK | OK | OK |    | -- | OK | OK | OK | OK |
  2.6.20 | OK | OK |    | #1 | OK | OK | OK | -- | OK | OK | OK | #1 |
  21-rc5 | OK | OK |    | OK | OK | OK | OK | -- | OK | OK | OK | OK |
  2.6.21 | OK | OK |    |    | OK | OK | OK | -- | OK | OK | OK |    |
  2.6.22 | OK | OK |    |    | OK | OK | OK | -- | OK | OK | OK |    |
  2.6.23 | OK | OK |    |    | OK | OK | OK | -- | OK | OK | OK |    |
  2.6.24 | OK | OK |    |    | OK | OK | OK | -- | OK | OK | OK |    |
  2.6.25 | OK | ** |    |    |    | ** | OK | -- | OK | OK | OK |    |
  2.6.26 | OK | ** |    |    |    | ** | OK | -- | OK | OK | OK |    |
  2.6.27 | OK | ** |    |    |    | ** | OK | -- | #2 | OK | OK |    |
  2.6.28 | OK | ** |    |    |    | ** | OK | -- | OK | OK | OK |    |
  2.6.29 | OK | ** |    |    |    | ** | OK | -- | OK | OK | OK |    |
  2.6.30 | OK | ** |    |    |    | ** | OK | -- | OK | OK | OK |    |
  2.6.31 | OK | ** |    |    |    | ** |    | -- | OK | OK | OK |    |
  2.6.32 | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    | OK
  2.6.33 | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  2.6.34 | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  2.6.35 | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  2.6.36 | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  2.6.37 | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  2.6.38 | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  2.6.39 | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  3.0    | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  3.1    | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  3.2    | OK | ** | OK |    |    | ** |    | -- | OK | OK |    |    |
  3.3    | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  3.4    | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  3.5    | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  3.6    | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  3.7    | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  3.8    | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  3.9    | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  3.10   | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  3.11   | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  3.12   | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  3.13   | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  3.14   | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  3.15   | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  3.16   | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  3.17   | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  3.18   | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  3.19   | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  4.0    | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  4.1    | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  4.2    | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  4.3    | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  4.4    | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  4.5    | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  4.6    | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  4.7    | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  4.8    | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  4.9    | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  4.10   | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  4.11   | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  4.12   | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  4.13   | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  4.14   | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  4.15   | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  4.16   | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  4.17   | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  4.18   | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  4.19   | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  4.20   | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  5.0    | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  5.1    | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |
  5.2    |    | ** |    |    |    | ** |    | -- |    | OK |    |    |
  5.3    |    | ** |    |    |    | ** |    | -- |    | OK |    |    |
  5.4    | OK | ** |    |    |    | ** |    | -- | OK | OK |    |    |

  OK    : Support.
  --    : Not support.
  Empty : Not test yet.
  TODO  : TODO.
  #1    : Both kexec/kdump does not work with PPC64 2.6.20 kernels. 
          This is a known problem fixed with later kernels.
  #2    : kdump does not work with i386 sparsemem 2.6.27 kernels. 
          This is a known problem fixed with later kernels.
  **    : The deleted memory model.
          On x86_64, both FLATMEM and DISCONTIGMEM have been deleted since
          linux-2.6.25.

* USAGE
  Please see "man makedumpfile" or "makedumpfile -h".

* TODO
  1. Supporting more kernels.
  2. Fixing the report message.
     - Now, it sometimes counts the number of free_pages duplicating zero-pages
       if creating an ELF dumpfile.
     - The cyclic mode will show incorrect and many progress indicator.

* NOTE
  1. A vmcoreinfo file should be generated by the makedumpfile which is used
     for dump filtering. If installing the latest makedumpfile, the vmcoreinfo
     file should be regenerated by it. Actually, makedumpfile v1.2.0 or later
     cannot work by vmcoreinfo file which is generated by v1.1.9 or before.

  2. If using linux-2.6.24 and kexec-tools-testing-20061214, /proc/vmcore
     contains vmcoreinfo data. So it is possible to create a dumpfile without
     -x option (or -i option) like the following:
     # makedumpfile -d 31 /proc/vmcore dumpfile

  3. On makedumpfile-1.2.5 or before, both '--xen-syms' and '--xen-vmcoreinfo'

debian/changelog
makedumpfile (1:1.6.7-1ubuntu2.3) focal; urgency=medium

  * makedumpfile falls back to cp with the 5.11 kernel (LP: #1942784)
    -d/p/0008-PATCH-Retrieve-MAX_PHYSMEM_BITS-from-vmcoreinfo.patch
  * makedumpfile doesn't create dmesg file on 5.10+ kernels (LP: #1921403)
    -d/p/0009-PATCH-1-2-printk-add-support-for-lockless-ringbuffer.patch
    -d/p/0010-PATCH-2-2-printk-use-committed-finalized-state-value.patch

 -- Ioanna Alifieraki <ioanna-maria.alifieraki@canonical.com>  Mon, 18 Oct 2021 16:43:56 +0000

makedumpfile (1:1.6.7-1ubuntu2.2) focal; urgency=medium

  * Fix makedumpfile failure on arm64 with 5.4 kernels (LP: #1879214)
    -d/p/0004-PATCH-arm64-Add-support-for-ARMv8.2-LPA-52-bit-PA-su.patch
    -d/p/0005-PATCH-1-3-Use-vmcoreinfo-note-in-proc-kcore-for-mem-.patch
    -d/p/0006-PATCH-2-3-arm64-Make-use-of-NUMBER-VA_BITS-in-vmcore.patch
    -d/p/0007-PATCH-3-3-arm64-support-flipped-VA-and-52-bit-kernel.patch

 -- Ioanna Alifieraki <ioanna-maria.alifieraki@canonical.com>  Thu, 15 Apr 2021 22:42:25 +0300

makedumpfile (1:1.6.7-1ubuntu2.1) focal; urgency=medium

  * d/p/0003-align_PMD_SECTION_MASK_with_PHYS_MASK.patch (LP: #1869465)
    Fix error on arm64 with 1TB+ of physical or firmware-mapped RAM.

 -- Ioanna Alifieraki <ioanna-maria.alifieraki@canonical.com>  Thu, 04 Jun 2020 14:47:17 +0100

makedumpfile (1:1.6.7-1ubuntu2) focal; urgency=low

  * debian/control/tests: sleep while waiting for systemd service.
    Also, exit in case the service is not successfully started, and dump
    status output for debugging.
    This is necessary because now that the initrd is not generated during
    postinst, it will be generated when the service is first started, and that
    may delay it for a few seconds.

 -- Thadeu Lima de Souza Cascardo <cascardo@canonical.com>  Tue, 17 Mar 2020 16:40:38 -0300

makedumpfile (1:1.6.7-1ubuntu1) focal; urgency=low

  [ Thadeu Lima de Souza Cascardo ]
  * Merge from Debian unstable.  Remaining changes:
    - Bump amd64 crashkernel from 384M-:128M to 512M-:192M.
    - Add a systemd-resolved service dependency in order kdump-tools is able
      to resolve DNS when in kdump boot.

  [ Dan Streetman ]
  * FeatureFreeze exception note:
    I checked upstream log from 1.6.6-1.6.7 and confirmed this is bugfix-only

 -- Dan Streetman <ddstreet@canonical.com>  Fri, 13 Mar 2020 10:25:32 -0400

makedumpfile (1:1.6.7-1) unstable; urgency=medium

  * Update to new upstream version 1.6.7.
  * Skip initrd generation during postinst. (Closes: #856594)

 -- Thadeu Lima de Souza Cascardo <cascardo@debian.org>  Mon, 02 Mar 2020 04:48:06 -0300

makedumpfile (1:1.6.6-4ubuntu3) focal; urgency=medium

  * No-change rebuild with fixed binutils on arm64.

 -- Matthias Klose <doko@ubuntu.com>  Sat, 08 Feb 2020 11:06:59 +0000

makedumpfile (1:1.6.6-4ubuntu2) focal; urgency=medium

  * x86_64: Fix an error due to makedumpfile being out-of-sync with recent
    kernels. To achieve that, add the following patch: "Increase
    SECTION_MAP_LAST_BIT to 4". (LP: #1857616)

 -- Guilherme G. Piccoli <gpiccoli@canonical.com>  Fri, 03 Jan 2020 16:57:00 -0300

makedumpfile (1:1.6.6-4ubuntu1) focal; urgency=medium

  [ Thadeu Lima de Souza Cascardo ]
  * Merge from Debian unstable.  Remaining changes:
    - Bump amd64 crashkernel from 384M-:128M to 512M-:192M.
  * Use reset_devices as a cmdline parameter. (LP: #1800566)
  * Use kdump-config reload after cpu or memory hotplug. (LP: #1828596)

  [ Guilherme G. Piccoli ]
  * Add a systemd-resolved service dependency in order kdump-tools is able
    to resolve DNS when in kdump boot. (LP: #1856323)

 -- Thadeu Lima de Souza Cascardo <cascardo@canonical.com>  Wed, 18 Dec 2019 14:38:51 -0300

makedumpfile (1:1.6.6-4) unstable; urgency=medium

  * Let the kernel decide the crashkernel offset for ppc64el (LP: #1741860)
  * kdump-config: implement try-reload
  * udev: hotplug: use try-reload
  * Set Rules-Requires-Root to no

 -- Thadeu Lima de Souza Cascardo <cascardo@debian.org>  Wed, 18 Dec 2019 10:50:47 -0300

makedumpfile (1:1.6.6-3) unstable; urgency=medium

  * Add a reload command.
  * Use kdump-config reload after cpu or memory hotplug.
  * Use reset_devices as a cmdline parameter.

 -- Thadeu Lima de Souza Cascardo <cascardo@debian.org>  Wed, 18 Sep 2019 01:00:16 +0000

makedumpfile (1:1.6.6-2ubuntu1) eoan; urgency=medium

  [ Thadeu Lima de Souza Cascardo ]
  * Merge from Debian unstable.  Remaining changes:

.pc/0009-PATCH-2-3-arm64-Make-use-of-NUMBER-VA_BITS-in-vmcore.patch/arch/arm64.c
/*
 * arch/arm64.c : Based on arch/arm.c
 *
 * Copyright (C) 2015 Red Hat, Pratyush Anand <panand@redhat.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation (version 2 of the License).
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#ifdef __aarch64__

#include "../elf_info.h"
#include "../makedumpfile.h"
#include "../print_info.h"

typedef struct {
	unsigned long pgd;
} pgd_t;

typedef struct {
	pgd_t pgd;
} pud_t;

typedef struct {
	pud_t pud;
} pmd_t;

typedef struct {
	unsigned long pte;
} pte_t;

#define __pte(x)	((pte_t) { (x) } )
#define __pmd(x)	((pmd_t) { (x) } )
#define __pud(x)	((pud_t) { (x) } )
#define __pgd(x)	((pgd_t) { (x) } )

static int lpa_52_bit_support_available;
static int pgtable_level;
static int va_bits;
static unsigned long kimage_voffset;

#define SZ_4K			4096
#define SZ_16K			16384
#define SZ_64K			65536

#define PAGE_OFFSET_36		((0xffffffffffffffffUL) << 36)
#define PAGE_OFFSET_39		((0xffffffffffffffffUL) << 39)
#define PAGE_OFFSET_42		((0xffffffffffffffffUL) << 42)
#define PAGE_OFFSET_47		((0xffffffffffffffffUL) << 47)
#define PAGE_OFFSET_48		((0xffffffffffffffffUL) << 48)
#define PAGE_OFFSET_52		((0xffffffffffffffffUL) << 52)

#define pgd_val(x)		((x).pgd)
#define pud_val(x)		(pgd_val((x).pgd))
#define pmd_val(x)		(pud_val((x).pud))
#define pte_val(x)		((x).pte)

/* See 'include/uapi/linux/const.h' for definitions below */
#define __AC(X,Y)	(X##Y)
#define _AC(X,Y)	__AC(X,Y)
#define _AT(T,X)	((T)(X))

/* See 'include/asm/pgtable-types.h' for definitions below */
typedef unsigned long pteval_t;
typedef unsigned long pmdval_t;
typedef unsigned long pudval_t;
typedef unsigned long pgdval_t;

#define PAGE_SHIFT	PAGESHIFT()

/* See 'arch/arm64/include/asm/pgtable-hwdef.h' for definitions below */

#define ARM64_HW_PGTABLE_LEVEL_SHIFT(n)	((PAGE_SHIFT - 3) * (4 - (n)) + 3)

#define PTRS_PER_PTE		(1 << (PAGE_SHIFT - 3))

/*
 * PMD_SHIFT determines the size a level 2 page table entry can map.
 */
#define PMD_SHIFT		ARM64_HW_PGTABLE_LEVEL_SHIFT(2)
#define PMD_SIZE		(_AC(1, UL) << PMD_SHIFT)
#define PMD_MASK		(~(PMD_SIZE-1))
#define PTRS_PER_PMD		PTRS_PER_PTE

/*
 * PUD_SHIFT determines the size a level 1 page table entry can map.
 */
#define PUD_SHIFT		ARM64_HW_PGTABLE_LEVEL_SHIFT(1)
#define PUD_SIZE		(_AC(1, UL) << PUD_SHIFT)
#define PUD_MASK		(~(PUD_SIZE-1))
#define PTRS_PER_PUD		PTRS_PER_PTE

/*
 * PGDIR_SHIFT determines the size a top-level page table entry can map
 * (depending on the configuration, this level can be 0, 1 or 2).
 */
#define PGDIR_SHIFT		ARM64_HW_PGTABLE_LEVEL_SHIFT(4 - (pgtable_level))
#define PGDIR_SIZE		(_AC(1, UL) << PGDIR_SHIFT)
#define PGDIR_MASK		(~(PGDIR_SIZE-1))
#define PTRS_PER_PGD		(1 << ((va_bits) - PGDIR_SHIFT))

/*
 * Section address mask and size definitions.
 */
#define SECTIONS_SIZE_BITS	30

/*
 * Hardware page table definitions.
 *
 * Level 1 descriptor (PUD).
 */
#define PUD_TYPE_TABLE		(_AT(pudval_t, 3) << 0)
#define PUD_TABLE_BIT		(_AT(pudval_t, 1) << 1)
#define PUD_TYPE_MASK		(_AT(pudval_t, 3) << 0)
#define PUD_TYPE_SECT		(_AT(pudval_t, 1) << 0)

/*
 * Level 2 descriptor (PMD).
 */
#define PMD_TYPE_MASK		(_AT(pmdval_t, 3) << 0)
#define PMD_TYPE_FAULT		(_AT(pmdval_t, 0) << 0)
#define PMD_TYPE_TABLE		(_AT(pmdval_t, 3) << 0)
#define PMD_TYPE_SECT		(_AT(pmdval_t, 1) << 0)
#define PMD_TABLE_BIT		(_AT(pmdval_t, 1) << 1)

/*
 * Level 3 descriptor (PTE).
 */
#define PTE_ADDR_LOW		(((_AT(pteval_t, 1) << (48 - PAGE_SHIFT)) - 1) << PAGE_SHIFT)
#define PTE_ADDR_HIGH		(_AT(pteval_t, 0xf) << 12)

static inline unsigned long
get_pte_addr_mask_arm64(void)
{
	if (lpa_52_bit_support_available)
		return (PTE_ADDR_LOW | PTE_ADDR_HIGH);
	else
		return PTE_ADDR_LOW;
}

#define PTE_ADDR_MASK		get_pte_addr_mask_arm64()

#define PAGE_MASK		(~(PAGESIZE() - 1))
#define PAGE_PRESENT		(1 << 0)

/* Helper API to convert between a physical address and its placement
 * in a page table entry, taking care of 52-bit addresses.
 */
static inline unsigned long
__pte_to_phys(pte_t pte)
{
	if (lpa_52_bit_support_available)
		return ((pte_val(pte) & PTE_ADDR_LOW) | ((pte_val(pte) & PTE_ADDR_HIGH) << 36));
	else
		return (pte_val(pte) & PTE_ADDR_MASK);
}

/* Find an entry in a page-table-directory */
#define pgd_index(vaddr)		(((vaddr) >> PGDIR_SHIFT) & (PTRS_PER_PGD - 1))

static inline pte_t
pgd_pte(pgd_t pgd)
{
	return __pte(pgd_val(pgd));
}

#define __pgd_to_phys(pgd)		__pte_to_phys(pgd_pte(pgd))
#define pgd_offset(pgd, vaddr)		((pgd_t *)(pgd) + pgd_index(vaddr))

static inline pte_t pud_pte(pud_t pud)
{
	return __pte(pud_val(pud));
}

static inline unsigned long
pgd_page_paddr(pgd_t pgd)
{
	return __pgd_to_phys(pgd);
}

/* Find an entry in the first-level page table. */
#define pud_index(vaddr)		(((vaddr) >> PUD_SHIFT) & (PTRS_PER_PUD - 1))
#define __pud_to_phys(pud)		__pte_to_phys(pud_pte(pud))

static inline unsigned long
pud_page_paddr(pud_t pud)
{
	return __pud_to_phys(pud);
}

/* Find an entry in the second-level page table. */
#define pmd_index(vaddr)		(((vaddr) >> PMD_SHIFT) & (PTRS_PER_PMD - 1))

static inline pte_t pmd_pte(pmd_t pmd)
{
	return __pte(pmd_val(pmd));
}

#define __pmd_to_phys(pmd)		__pte_to_phys(pmd_pte(pmd))

static inline unsigned long
pmd_page_paddr(pmd_t pmd)
{
	return __pmd_to_phys(pmd);
}

/* Find an entry in the third-level page table. */
#define pte_index(vaddr)		(((vaddr) >> PAGESHIFT()) & (PTRS_PER_PTE - 1))
#define pte_offset(dir, vaddr)		(pmd_page_paddr((*dir)) + pte_index(vaddr) * sizeof(pte_t))

static unsigned long long
__pa(unsigned long vaddr)
{
	if (kimage_voffset == NOT_FOUND_NUMBER ||
			(vaddr >= PAGE_OFFSET))
		return (vaddr - PAGE_OFFSET + info->phys_base);
	else
		return (vaddr - kimage_voffset);
}

static pud_t *
pud_offset(pgd_t *pgda, pgd_t *pgdv, unsigned long vaddr)
{
	if (pgtable_level > 3)
		return (pud_t *)(pgd_page_paddr(*pgdv) + pud_index(vaddr) * sizeof(pud_t));
	else
		return (pud_t *)(pgda);
}

static pmd_t *
pmd_offset(pud_t *puda, pud_t *pudv, unsigned long vaddr)
{
	if (pgtable_level > 2)
		return (pmd_t *)(pud_page_paddr(*pudv) + pmd_index(vaddr) * sizeof(pmd_t));
	else
		return (pmd_t *)(puda);
}

static int calculate_plat_config(void)
{
	/* derive pgtable_level as per arch/arm64/Kconfig */
	if ((PAGESIZE() == SZ_16K && va_bits == 36) ||
			(PAGESIZE() == SZ_64K && va_bits == 42)) {
		pgtable_level = 2;
	} else if ((PAGESIZE() == SZ_64K && va_bits == 48) ||
			(PAGESIZE() == SZ_4K && va_bits == 39) ||
			(PAGESIZE() == SZ_16K && va_bits == 47)) {
		pgtable_level = 3;
	} else if ((PAGESIZE() != SZ_64K && va_bits == 48)) {
		pgtable_level = 4;
	} else {
		ERRMSG("PAGE SIZE %#lx and VA Bits %d not supported\n",
				PAGESIZE(), va_bits);
		return FALSE;
	}

	return TRUE;
}

unsigned long
get_kvbase_arm64(void)
{
	return (0xffffffffffffffffUL << va_bits);
}

int
get_phys_base_arm64(void)
{
	int i;
	unsigned long long phys_start;
	unsigned long long virt_start;

	if (NUMBER(PHYS_OFFSET) != NOT_FOUND_NUMBER) {
		info->phys_base = NUMBER(PHYS_OFFSET);
		DEBUG_MSG("phys_base    : %lx (vmcoreinfo)\n",
				info->phys_base);
		return TRUE;
	}

	if (get_num_pt_loads() && PAGE_OFFSET) {
		for (i = 0;
		    get_pt_load(i, &phys_start, NULL, &virt_start, NULL);
		    i++) {
			if (virt_start != NOT_KV_ADDR
			    && virt_start >= PAGE_OFFSET
			    && phys_start != NOT_PADDR) {
				info->phys_base = phys_start -
					(virt_start & ~PAGE_OFFSET);
				DEBUG_MSG("phys_base    : %lx (pt_load)\n",
						info->phys_base);
				return TRUE;
			}
		}
	}

	ERRMSG("Cannot determine phys_base\n");
	return FALSE;
}

unsigned long
get_kaslr_offset_arm64(unsigned long vaddr)
{
	unsigned int i;
	char buf[BUFSIZE_FGETS], *endp;

	if (!info->kaslr_offset && info->file_vmcoreinfo) {
		if (fseek(info->file_vmcoreinfo, 0, SEEK_SET) < 0) {
			ERRMSG("Can't seek the vmcoreinfo file(%s). %s\n",
					info->name_vmcoreinfo, strerror(errno));
			return FALSE;
		}

		while (fgets(buf, BUFSIZE_FGETS, info->file_vmcoreinfo)) {
			i = strlen(buf);
			if (!i)
				break;
			if (buf[i - 1] == '\n')
				buf[i - 1] = '\0';
			if (strncmp(buf, STR_KERNELOFFSET,
					strlen(STR_KERNELOFFSET)) == 0) {
				info->kaslr_offset =
					strtoul(buf+strlen(STR_KERNELOFFSET),&endp,16);
				DEBUG_MSG("info->kaslr_offset: %lx\n", info->kaslr_offset);
			}
		}
	}

	if (vaddr >= __START_KERNEL_map &&
			vaddr < __START_KERNEL_map + info->kaslr_offset) {
		DEBUG_MSG("info->kaslr_offset: %lx\n", info->kaslr_offset);
		return info->kaslr_offset;
	} else {
		/*
		 * TODO: we need to check if it is vmalloc/vmmemmap/module
		 * address, we will have different offset
		 */
		return 0;
	}
}

ulong
get_stext_symbol(void)
{
	int found;
	FILE *fp;
	char buf[BUFSIZE];
	char *kallsyms[MAXARGS];
	ulong kallsym;

	if (!file_exists("/proc/kallsyms")) {
		ERRMSG("(%s) does not exist, will not be able to read symbols. %s\n",
		       "/proc/kallsyms", strerror(errno));
		return FALSE;
	}

	if ((fp = fopen("/proc/kallsyms", "r")) == NULL) {
		ERRMSG("Cannot open (%s) to read symbols. %s\n",
		       "/proc/kallsyms", strerror(errno));
		return FALSE;
	}

	found = FALSE;
	kallsym = 0;

	while (!found && fgets(buf, BUFSIZE, fp) &&
	      (parse_line(buf, kallsyms) == 3)) {
		if (hexadecimal(kallsyms[0], 0) &&
		    STREQ(kallsyms[2], "_stext")) {
			kallsym = htol(kallsyms[0], 0);
			found = TRUE;
			break;
		}
	}
	fclose(fp);

	return(found ? kallsym : FALSE);
}

int
get_machdep_info_arm64(void)
{
	/* Determine if the PA address range is 52-bits: ARMv8.2-LPA */
	if (NUMBER(MAX_PHYSMEM_BITS) != NOT_FOUND_NUMBER) {
		info->max_physmem_bits = NUMBER(MAX_PHYSMEM_BITS);
		if (info->max_physmem_bits == 52)
			lpa_52_bit_support_available = 1;
	} else
		info->max_physmem_bits = 48;

	/* Check if va_bits is still not initialized. If still 0, call
	 * get_versiondep_info() to initialize the same.
	 */
	if (!va_bits)
		get_versiondep_info_arm64();

	if (!calculate_plat_config()) {
		ERRMSG("Can't determine platform config values\n");
		return FALSE;
	}

	kimage_voffset = NUMBER(kimage_voffset);
	info->section_size_bits = SECTIONS_SIZE_BITS;

	DEBUG_MSG("kimage_voffset   : %lx\n", kimage_voffset);
	DEBUG_MSG("max_physmem_bits : %ld\n", info->max_physmem_bits);
	DEBUG_MSG("section_size_bits: %ld\n", info->section_size_bits);

	return TRUE;
}

unsigned long long
kvtop_xen_arm64(unsigned long kvaddr)
{
	return ERROR;
}

int
get_xen_basic_info_arm64(void)
{
	return ERROR;
}

int
get_xen_info_arm64(void)
{
	return ERROR;
}

int
get_versiondep_info_arm64(void)
{
	ulong _stext;

	_stext = get_stext_symbol();
	if (!_stext) {
		ERRMSG("Can't get the symbol of _stext.\n");
		return FALSE;
	}

	/* Derive va_bits as per arch/arm64/Kconfig */
	if ((_stext & PAGE_OFFSET_36) == PAGE_OFFSET_36) {
		va_bits = 36;
	} else if ((_stext & PAGE_OFFSET_39) == PAGE_OFFSET_39) {
		va_bits = 39;
	} else if ((_stext & PAGE_OFFSET_42) == PAGE_OFFSET_42) {
		va_bits = 42;
	} else if ((_stext & PAGE_OFFSET_47) == PAGE_OFFSET_47) {
		va_bits = 47;
	} else if ((_stext & PAGE_OFFSET_48) == PAGE_OFFSET_48) {
		va_bits = 48;
	} else {
		ERRMSG("Cannot find a proper _stext for calculating VA_BITS\n");
		return FALSE;
	}

	info->page_offset = (0xffffffffffffffffUL) << (va_bits - 1);

	DEBUG_MSG("va_bits      : %d\n", va_bits);
	DEBUG_MSG("page_offset  : %lx\n", info->page_offset);

	return TRUE;
}

/* 1GB section for Page Table level = 4 and Page Size = 4KB */
static int
is_pud_sect(pud_t pud)
{
	return ((pud_val(pud) & PUD_TYPE_MASK) == PUD_TYPE_SECT);
}

static int
is_pmd_sect(pmd_t pmd)
{
	return ((pmd_val(pmd) & PMD_TYPE_MASK) == PMD_TYPE_SECT);
}

/*
 * vaddr_to_paddr_arm64() - translate arbitrary virtual address to physical
 * @vaddr: virtual address to translate
 *
 * Function translates @vaddr into physical address using page tables. This
 * address can be any virtual address. Returns physical address of the
 * corresponding virtual address or %NOT_PADDR when there is no translation.
 */
unsigned long long
vaddr_to_paddr_arm64(unsigned long vaddr)
{
	unsigned long long paddr = NOT_PADDR;
	unsigned long long swapper_phys;
	pgd_t	*pgda, pgdv;
	pud_t	*puda, pudv;
	pmd_t	*pmda, pmdv;
	pte_t 	*ptea, ptev;

	if (SYMBOL(swapper_pg_dir) == NOT_FOUND_SYMBOL) {
		ERRMSG("Can't get the symbol of swapper_pg_dir.\n");
		return NOT_PADDR;
	}

	swapper_phys = __pa(SYMBOL(swapper_pg_dir));

	pgda = pgd_offset(swapper_phys, vaddr);
	if (!readmem(PADDR, (unsigned long long)pgda, &pgdv, sizeof(pgdv))) {
		ERRMSG("Can't read pgd\n");
		return NOT_PADDR;
	}

	puda = pud_offset(pgda, &pgdv, vaddr);
	if (!readmem(PADDR, (unsigned long long)puda, &pudv, sizeof(pudv))) {
		ERRMSG("Can't read pud\n");
		return NOT_PADDR;
	}

	if (is_pud_sect(pudv)) {
		paddr = (pud_page_paddr(pudv) & PUD_MASK) +
				(vaddr & (PUD_SIZE - 1));
		return paddr;
	}

	pmda = pmd_offset(puda, &pudv, vaddr);
	if (!readmem(PADDR, (unsigned long long)pmda, &pmdv, sizeof(pmdv))) {
		ERRMSG("Can't read pmd\n");
		return NOT_PADDR;
	}

	if (is_pmd_sect(pmdv)) {
		paddr = (pmd_page_paddr(pmdv) & PMD_MASK) +
				(vaddr & (PMD_SIZE - 1));
		return paddr;
	}

	ptea = (pte_t *)pte_offset(&pmdv, vaddr);
	if (!readmem(PADDR, (unsigned long long)ptea, &ptev, sizeof(ptev))) {
		ERRMSG("Can't read pte\n");
		return NOT_PADDR;
	}

	if (!(pte_val(ptev) & PAGE_PRESENT)) {
		ERRMSG("Can't get a valid pte.\n");
		return NOT_PADDR;
	} else {
		paddr = __pte_to_phys(ptev) +
				(vaddr & (PAGESIZE() - 1));
	}

	return paddr;
}

#endif /* __aarch64__ */
